datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -450,6 +463,9 @@ impl<'a> TreeNodeRewriter for ConstEvaluator<'a> {
                     ConstSimplifyResult::Simplified(s) => {
                         Ok(Transformed::yes(Expr::Literal(s)))
                     }
+                    ConstSimplifyResult::NotSimplified(s) => {
```

Comment:
I had to add a new variant to this result type so that the constant evaluator would return with `Transformed::no` when it encounters a `Literal`. It was always returning `Transformed::yes` and forcing the loop to go to the max iterations.

Reply:
I think we can define `NotSimplified` without value, so we might not accidentally return a "simplified/rewritten" expr back

```rust
enum ConstSimplifyResult {
    // Expr was simplifed and contains the new expression
    Simplified(ScalarValue),
    // Expr was not simplified and original value is returned
    NotSimplified,
    // Evaluation encountered an error, contains the original expression
    SimplifyRuntimeError(DataFusionError, Expr),
}


ConstSimplifyResult::NotSimplified => {
   Ok(Transformed::no(expr))
}
```

Reply:
As long as we can avoid copying the Expr then this seems like a good idea to me ❤️